### PR TITLE
 Fix for #18178 and #18187 by changing the concat of empty RangeIndex 

### LINKED ISSUE #18178: BUG: pd.concat raises if called on mixture of empty and non-empty dataframes
I noticed a change in how pd.concat works between 0.20.3 and 0.21.0:

```python
import pandas as pd

df1 = pd.DataFrame({'foo': [1]})
df2 = pd.DataFrame({'foo': []})

res = pd.concat([df1, df2])
```

This example does not raise an exception in 0.20.3. In 0.21.0, it raises the following error:

```
Traceback (most recent call last):
  File "test.py", line 6, in <module>
    res = pd.concat([df1, df2])
  File "/homes/joewag/miniconda3/envs/py3/lib/python3.6/site-packages/pandas/core/reshape/concat.py", line 213, in concat
    return op.get_result()
  File "/homes/joewag/miniconda3/envs/py3/lib/python3.6/site-packages/pandas/core/reshape/concat.py", line 408, in get_result
    copy=self.copy)
  File "/homes/joewag/miniconda3/envs/py3/lib/python3.6/site-packages/pandas/core/internals.py", line 5202, in concatenate_block_managers
    return BlockManager(blocks, axes)
  File "/homes/joewag/miniconda3/envs/py3/lib/python3.6/site-packages/pandas/core/internals.py", line 3028, in __init__
    self._verify_integrity()
  File "/homes/joewag/miniconda3/envs/py3/lib/python3.6/site-packages/pandas/core/internals.py", line 3239, in _verify_integrity
    construction_error(tot_items, block.shape[1:], self.axes)
  File "/homes/joewag/miniconda3/envs/py3/lib/python3.6/site-packages/pandas/core/internals.py", line 4603, in construction_error
    passed, implied))
ValueError: Shape of passed values is (1, 1), indices imply (1, 0)
```

#### Expected Output
I would expect no error, and all(res==df1) to be true

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-696.1.1.el6.centos.plus.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0
pytest: 3.2.3
pip: 9.0.1
setuptools: 36.2.2.post20170724
Cython: 0.26
numpy: 1.13.3
scipy: 0.19.1
pyarrow: None
xarray: None
IPython: 6.1.0
sphinx: 1.6.4
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: 3.4.0
numexpr: 2.6.2
feather: None
matplotlib: None
openpyxl: None
xlrd: 1.1.0
xlwt: None
xlsxwriter: 1.0.2
lxml: None
bs4: 4.6.0
html5lib: None
sqlalchemy: 1.1.14
pymysql: 0.7.11.None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #18187: BUG: pd.concat returns empty series if called on mixture of empty and non-empty series
#### Code Sample

```python
import pandas as pd

s1 = pd.Series([1])
s2 = pd.Series([])

pd.concat([s1, s2])
```
returns an empty series `Series([], dtype: int64)`.

#### Problem description

Concatenating an non-empty and empty series returns an empty series.
If this is supposed to be the intended behavior, it is very counterintuitive. I would rather expect that the
result of `pd.concat([s1, s2])` is simply equal to `s1`. In fact, this used to be the behavior in older versions like 0.20.3.

Btw, I stumbled upon this error while I was trying to fix #18178. Out of curiosity I checked what happens in case of concatenating non-empty and empty series instead of frames.

#### Expected Output

Series equal to `s1`, i.e. `repr(pd.concat([s1, s2]))` returns `'0    1\ndtype: int64'`.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-98-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0
pytest: None
pip: 9.0.1
setuptools: 36.5.0.post20170921
Cython: None
numpy: 1.13.3
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by SmokinCaterpillar**
The `_concat_rangeindex_same_dtype` function now keeps track of the last non-empty `RangeIndex` to extract the new stop value.

This fixes two issues with concatenating non-empty and empty `DataFrames` and `Series`.

Two regression tests were added as well.

- closes #18178 and closes #18187 
- 2 regression tests added
- Submission passes `git diff master -u -- "*.py" | flake8 --diff`
- 1 whatsnew entry

### CODE REVIEW

**Comment by SmokinCaterpillar** [comment]
As an alternative solution, instead of using
```python
if not len(obj):
    continue
```
and this new `last_non_empty` variable,
one could simply replace `continue` as follows:
```python
if not len(obj):
    from pandas import Int64Index
    return _concat_index_same_dtype(indexes, klass=Int64Index)
```
However, in this case, the returned new index would always be changed to an integer one in case of a single empty `RangeIndex` in `indexes`. Accordingly, some already existing tests would need to be changed as well.

**Comment by jreback** [comment]
i am ok with the soln

File: pandas/core/dtypes/concat.py [hunk 1]
```diff
@@ -572,11 +572,13 @@ def _concat_rangeindex_same_dtype(indexes):
...
     """
 
-    start = step = next = None
+    start = step = next = last_non_empty = None
 
     for obj in indexes:
         if not len(obj):
             continue
+        # Remember the last non-empty index for the stop value
+        last_non_empty = obj
```
Comment (jreback) [inline pandas/core/dtypes/concat.py]: maybe filter out he non empties first? then i think this loop should be ok?

**Comment by SmokinCaterpillar** [comment]
I made the code a bit nicer by first filtering the empty indexes with a list comprehension

**Review by jorisvandenbossche — APPROVED** [review]
Thanks!

Side note: the behaviour in case of all empties to take the start/stop/step feels a bit strange to me (why not the first? or convert to empty Int64Index?), but given that this is how it already was before, we can leave that for another PR (if we would want to change that)

**Comment by SmokinCaterpillar** [comment]
@jorisvandenbossche  yes, another pull request for these changes is probably a good idea, especially because IMO this involves changing existing RangeIndex tests.

**Comment by jreback** [comment]
thanks @SmokinCaterpillar 

yeah I find this logic slightly strange. If you'd like to follow up with a PR to clarify would be great!

**Comment by SmokinCaterpillar** [comment]
Opened a follow up PR #18214 :-)

**Pull request merged**